### PR TITLE
Adding dynamic memory management features (EDMM) with SGX2 supports 

### PR DESCRIPTION
**Pull request by rainfld**
This PR added three major new features which are supported by SGX2 EDMM instructions, including:
- Dynamic Heap Management. Graphene-SGX allocates EPC pages for the heap when the heap grows with the application is executed.
- Dynamic Stack Management. Same as heap, Graphene-SGX dynamically grows the stacks with EPC pages at runtime
- Dynamic Thread Management. Allocating necessary EPC memory for creating new threads at runtime

---
This change is [Reviewable](https://reviewable.io/reviews/oscarlab/graphene/234)

### CODE REVIEW

**Comment by rainfld** [comment]
Test this please

**Comment by chiache** [comment]
Do we need a SGX2 machine to test this patch?

**Comment by donporter** [comment]
I'll get a NUC added to our CI setup as soon as possible to test SGX v2 support.

**Comment by donporter** [comment]
Ok to test

**Comment by donporter** [comment]
BTW, the other thing that is missing here are some unit tests for this functionality.

**Comment by dimakuv** [comment]
The EDMM feature will be developed by @vijaydhanraj.

**Comment by johnvenn** [comment]
Per my understanding, this feature requires driver change to support those Intel instructions to expand or trim EPC pages, but now the in-kernel(DCAP) driver doesn't support EDMM. Do you mean this feature can only work with legacy OOT driver?

**Comment by vijaydhanraj** [comment]
For our initial development, we are evaluating with the legacy driver.

**Comment by mkow** [comment]
> Per my understanding, this feature requires driver change to support those Intel instructions to expand or trim EPC pages, but now the in-kernel(DCAP) driver doesn't support EDMM. Do you mean this feature can only work with legacy OOT driver?

Yes. The upstreamed in-tree driver also doesn't support it, but AFAIK there are plans to add this feature to it.

File: Pal/src/host/Linux-SGX/sgx_framework.c [hunk 1]
```diff
@@ -297,6 +297,27 @@ int add_pages_to_enclave(sgx_arch_secs_t * secs,
...
+    params.nr_pages = 1;
+    int ret = 0;
+
+    ret = INLINE_SYSCALL(ioctl, 3, isgx_device, SGX_IOC_ENCLAVE_MKTCS, &params);
+    if (IS_ERR(ret)) {
+	SGX_DBG(DBG_I, "Enclave MKTCS returned %d\n", ret);
+	return ;
+    }
+#else
+    SGX_DBG(DBG_E, "EDMM is not supported by SDK before 2.0\n");
```
Comment (duanbing) [inline Pal/src/host/Linux-SGX/sgx_framework.c:317]: I think that's not SDK, it's SGX driver?  https://github.com/intel/linux-sgx-driver/blob/master/sgx_ioctl.c#L383.

Comment (dimakuv) [reply Pal/src/host/Linux-SGX/sgx_framework.c:317]: @duanbing Yes, you are correct. But this PR is 3 years old and not planned for merging. We keep this PR open simply to not forget about a new version of the EDMM support :)

Comment (mkow) [reply Pal/src/host/Linux-SGX/sgx_framework.c:317]: @vijaydhanraj is reviving these efforts, if you're interested then please take a look at #2190.

**Comment by mkow** [comment]
Superseded by #2190, closing.

**Pull request closed without merging**